runtime/libia2/include/ia2.h

```diff
@@ -59,9 +62,10 @@
 ///
 /// Creates a new function with `ia2_sighandler_` prepended to the given
 /// function name which should be registered with sigaction().
-#define IA2_DEFINE_SIGACTION(function, pkey)                                   \
-  void ia2_sighandler_##function(int, siginfo_t *, void *);                    \
-  _IA2_DEFINE_SIGNAL_HANDLER(function, pkey)
+#define IA2_DEFINE_SIGACTION(function, pkey)                                        \
+  __attribute__((naked)) void ia2_sighandler_##function(int, siginfo_t *, void *) { \
+  _IA2_DEFINE_SIGNAL_HANDLER(function, pkey)                                        \
+  }
```

Comment:
If it's defined as a normal, albeit naked, function, I think the nested (not closure) function works fine.